File: Project.java
import java.time.*;
import java.util.*;

/**Project class that accesses data related to the
 * project, like its ID, name, employees, deadline,
 * and tasks.
 */
public class Project extends Activity {
    private List<Integer> derivedTaskIDs;

    //Constructors

    public Project() {
        this.setName("Unnamed Project");
    }

    public Project(String newProjectName) {
        this.setName(newProjectName);
    }

    public Project(String newProjectName, LocalDate newDeadline) {
        super(newProjectName, newDeadline);
    }

    //Getters and setters

    /**
     * @param newTaskIDs An ordered list of the tasks that are derived
     * from this project
     */
    public void setSubactivitys(List<Integer> newTaskIDs) {
        this.derivedTaskIDs = newTaskIDs;
    }

    /**
     * @return derivedTaskIDs The ordered list of tasks deriving from this project
     */
    public List<Integer> getDerivedTasks() {
        return this.derivedTaskIDs;
    }
}

File: models/project.js
// Assuming Activity is another class that exists in your project
class Activity {
    constructor(name, deadline) {
        this.name = name || "Unnamed Activity";
        this.deadline = deadline || new Date(); // Default to the current date
    }

    // Getter and setter for the name
    getName() {
        return this.name;
    }

    setName(newName) {
        this.name = newName;
    }

    // Getter and setter for the deadline
    getDeadline() {
        return this.deadline;
    }

    setDeadline(newDeadline) {
        this.deadline = new Date(newDeadline); // Convert to Date object
    }
}

class Project extends Activity {
    constructor(newProjectName, newDeadline) {
        super(newProjectName, newDeadline);
        this.derivedTaskIDs = [];
    }

    // Setter for derived task IDs
    setSubactivities(newTaskIDs) {
        this.derivedTaskIDs = Array.isArray(newTaskIDs) ? newTaskIDs : [];
    }

    // Getter for derived task IDs
    getDerivedTasks() {
        return this.derivedTaskIDs;
    }
}

// Example usage:

// Create a project with a name and deadline
const myProject = new Project("My New Project", "2024-12-31");

// Set some derived tasks
myProject.setSubactivities([1, 2, 3, 4]);

// Get the derived tasks
console.log(myProject.getDerivedTasks()); // [1, 2, 3, 4]
